### PR TITLE
Create stabilization_checklist.md

### LINKED ISSUE #3560: Write down the checklist for stabilizing a component
This has been on my to-do list for a very very long time but I keep failing to do it because there is no issue tracking it.

When stabilizing a component, we need to ensure that:

1. Style guide is followed (this includes a panic-free coding style)
2. FFI is complete
3. Docs are complete (including crate-level examples and coverage for all interesting options)
4. Data story is complete (data is highly modular, zero-copy, makes use of canonical ZeroVec best practices, and any necessary features added to datagen)
5. Crate exhibits i18n Correctness for all implemented APIs (no known deviation from Unicode specs) -- this is not the same as feature coverage
6. Final sign-off from an i18n expert who isn't on the core ICU4X team (usually Markus)

### PR DESCRIPTION
**Pull request by sffc**
Fixes #3560

### CODE REVIEW

**Review by robertbastian — CHANGES_REQUESTED** [review]
https://github.com/unicode-org/icu4x/pull/3397?

File: docs/process/stabilization_checklist.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
...
+
+1. [ ] **Style guide is followed.** To enforce the style guide, see [boilerplate.md](boilerplate.md).
+2. [ ] **FFI is complete.** There should be no unjustified suppressions nor entries in missing_apis.txt.
+3. [ ] **Docs are complete.** In addition to covering every exported function, there needs to be a crate-level overview example, and all options should be covered by at least one docs test.
+4. [ ] **Data story is complete.** The code and data should be highly modular, such that users do not need to carry more code or data than they need. The data should be zero-copy of course, and it should make use of the abstractions available in the zerovec crate. If datagen requires a new feature in order for the new data to be modularized, that feature should be implemented.
+5. [ ] **Feature exhibits i18n correctness.** There should be no known gaps in localization quality, meaning that for the component in question, a user in any CLDR locale should receive an experience on par with that of a user in English. In other words, the component does not need to be feature-complete, but of the features that are supported, they should be fully implemented.
```
Comment (Manishearth) [inline docs/process/stabilization_checklist.md:9]: Hm, I think this is rather strict and potentially ill-defined. For example, as Markus has mentioned in the context of https://github.com/unicode-org/icu4x/issues/3552 , that was not a feature ICU4C had for a long time: it's a relatively niche thing that required a bunch of work to design and implement, and was done after the fact.

In general I think ICU4X is going to have bugs and feature gaps. We have some known ones in segmentation as well because we implement spec rules rather than CLDR ones. I would consider that a gap in localization quality as well: the CLDR rules are known to be better than the spec (and the spec shall soon update to them). But segmentation is stable.

Whether or not a gap of this kind is a stabilization blocker, in this policy, largely depends on whether or not it's discovered before stabilization. If I had not tried the niche Greek ICU4C tests I would not have noticed this until later, since I did not have plans to import ICU4C's entire testsuite by hand.

I think a better rule is that each such feature gap is evaluated on a per-case basis, and, importantly, we ensure we are not blocked out of implementing it in the future. (Ideally, we document the gap as well)

**Review by Manishearth — CHANGES_REQUESTED** [review]
yeah we should also copy in all of the #3397 bullet points and organize appropriately

**Comment by sffc** [comment]
> #3397?

😆 

I'll pick one of the two PRs and pull the best parts of each one into the other

File: docs/process/stabilization_checklist.md (hunk 1 above)
Comment (sffc) [reply docs/process/stabilization_checklist.md:9]: The type of situation I would like to avoid is one where half-baked (or even 80%-baked) components spawn bugs that sit around for years with no one motivated to fix them. The stabilization of a component is the _only_ tool that ICU4X The Open Source Project holds to put real pressure on getting things fixed. After something ships, follow-up bugs just go in the pile of backlog issues and have a high likelihood of never getting fixed.

I'm not as happy as I would like to be about the state of the Segmenter component, as you know, but business pressure added up quarter after quarter and we decided to ship the best that the team could achieve. I really don't want to use the state of Segmenter as a precedent for what happens in other components.

Comment (sffc) [reply docs/process/stabilization_checklist.md:9]: > I actually don't agree on this: I think client needs are a pretty strong force here as well. I think we would be very careful selling something without ICU4C parity when it coems to i18n-completeness to a production user: for example, we cannot recommend icu_casemap to browser authors as long as the Greek thing isn't fixed. But I don't think that ought to block stabilization: There are plenty of clients for whom this may not matter at all. (e.g. it does not matter for Flutter).

So in other words, the other lever is parity with ICU4C. It is rare for a client to put any amount of resourcing behind fixing fix an i18n correctness issue _for the sake of i18n correctness_.

Comment (Manishearth) [reply docs/process/stabilization_checklist.md:9]: My assertion here is that when it comes to a lot of this stuff, "parity with ICU4C" is not materially different from "i18n correct", because there is not One True Answer for many of these things and instead _someone_ has to set an industry standard as to what these operations mean in a pan-linguistic way. Sometimes that is the Unicode standard, sometimes it is ICU4C. Hopefully we may be the ones blazing those frontiers as well in the future as we start picking up things not covered by ICU4C.

This is not the case for _all_ of this stuff. It is the case for akshara segmentation and I would argue Greek casemapping as well (bear in mind: ICU4C's Greek casemapping behavior _is a divergence from the spec_!). There are absolutely some things where there is One True Answer. It's not always obvious when that is the case. 

I do think that's the crux of it: There's different types of i18n correctness, some of them are absolute and unchanging, whereas some of them are a carefully weighed tradeoff which may come up at any time. I would absolutely be in support of a rule that prevented us from shipping e.g. plural rule support that couldn't handle 5 or 6 plural forms, or decimal formatting support that doesn't handle systems where the separators are not evenly spaced apart. But I would be much more reluctant for us to do this for akshara segmentation or Greek casing which are pretty nuanced issues with multiple potential positions. (I would still very much be in support of us documenting our lack of support for those). Because fundamentally, the latter kind can crop up any time, and we are not guaranteed to know all or even most of them before stabilization.

----

(I picked browsers as an example because it's a case where the Web has collectively decided that ICU4C's behavior is standard. This is de-facto, not de-jure, since the [spec](https://drafts.csswg.org/css-text/#text-transform-mapping) mostly says "do what the Unicode standard does, but you can also do extra per-language stuff".)

**Comment by sffc** [comment]
Closing in favor of #3397
